src/Vault.sol

```diff
@@ -0,0 +1,166 @@
+
+    ///@dev Emitted on liquidation()
+    ///@param collateralAsset The address of the debt asset
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@dev To set collateralization level of debt engine
+    ///@notice To allow for margin call at defined level, preventing Vault from bearing risk in fast moving markets
+    uint public immutable bufferDenominator; 
+    uint public immutable bufferNumerator;
+
+    ///@dev Returns decimal places of asset prices as dictated by Chainlink Oracle
+    uint public immutable decimals;
+    
+    constructor(address dai_, address weth_, address priceFeedAddress, uint bufferNumerator_, uint bufferDenominator_) {
+        collateral = IERC20(weth_);
+        debt = IERC20(dai_);
+
+        priceFeed = AggregatorV3Interface(priceFeedAddress);
+        decimals = priceFeed.decimals();
+
+        // collateralization level
+        bufferNumerator = bufferNumerator_;
+        bufferDenominator = bufferDenominator_;
+    }
+
+    ///@dev Users deposit collateral asset into Vault
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for margined colleteralization colculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
+        uint maxDebt = getMaxDebt(msg.sender);
+        require(debtAmount <= maxDebt, "Insufficient collateral!");
+
+        debts[msg.sender] += debtAmount;
+        bool sent = debt.transfer(msg.sender, debtAmount);
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Get maximum debt asset borrowable, calculated based on collateralization level
+    ///@dev Collateralization level: bufferNumerator/bufferDenominator (e.g. 8/10 -> 80%) | Think of this as margining
+    ///@param user User address to calculate maximum debt possible based on collateralization level
+    function getMaxDebt(address user) public view returns(uint) {
+        uint availableCollateral = (deposits[user]/bufferDenominator)*bufferNumerator;
```

Comment:
get in the habit of multiplying first then dividing after, this wei u wont lose precision

Reply:
lol nice pun - yup its something i need to be concussions about

Reply:
removed this function as advised by Alberto